docs/reference/scripting.asciidoc

```diff
@@ -4,61 +4,72 @@
-to specify the language of the script.
+The default scripting language is <<modules-scripting-painless,Painless>>.
+Additional `lang` plugins are available to run scripts written in other
+languages. You can specify the language of the script anywhere that scripts run.
 
 [discrete]
-== General-purpose languages
+[[scripts-and-search-speed]]
+== Scripts and search speed
 
-These languages can be used for any purpose in the scripting APIs,
-and give the most flexibility.
+Scripts can't use {es}'s index structures or related optimizations, which can
+sometimes result in slower search speeds.
 
-[cols="<,<,<",options="header",]
-|=======================================================================
-|Language
-    |Sandboxed
-    |Required plugin
+If you often use scripts to transform indexed data, you can speed up search by
+making these changes during ingest instead. However, that often means slower
+index speeds.
 
-|<<modules-scripting-painless, `painless`>>
-    |yes
-    |built-in
+[discrete]
+[[modules-scripting-using-caching]]
+== Script caching
+
+All scripts are cached by default so that they only need to be recompiled
+when updates occur. By default, scripts do not have a time-based expiration.
+You can change this behavior by using the `script.cache.expire` setting.
+Use the `script.cache.max_size` setting to configure the size of the cache.
 
-|=======================================================================
+NOTE: The size of scripts is limited to 65,535 bytes. Set the value of `script.max_size_in_bytes` to increase that soft limit. If your scripts are
+really large, then consider using a
+<<modules-scripting-engine,native script engine>>.
 
 [discrete]
-== Special-purpose languages
+[[scripting-available-languages]]
+== Available scripting languages
 
-These languages are less flexible, but typically have higher performance for
-certain tasks.
+Painless is purpose-built for {es}, can be used for any purpose in the
+scripting APIs, and provides the most flexibility. The other languages are less
+flexible, but typically have higher performance for certain tasks.
 
 [cols="<,<,<,<",options="header",]
-|=======================================================================
+|========
 |Language
     |Sandboxed
     |Required plugin
     |Purpose
 
+|<<modules-scripting-painless,`painless`>>
+    |{yes-icon}
+    |Built-in
+    |Purpose-built for {es}
+
 |<<modules-scripting-expression, `expression`>>
-    |yes
-    |built-in
-    |fast custom ranking and sorting
+    |{yes-icon}
+    |Built-in
+    |Fast custom ranking and sorting
 
 |<<search-template, `mustache`>>
-    |yes
-    |built-in
-    |templates
+    |{yes-icon}
+    |Built-in
+    |Templates
 
 |<<modules-scripting-engine, `java`>>
-    |n/a
-    |you write it!
-    |expert API
-
-|=======================================================================
+    |{no-icon}
+    |You write it!
+    |Expert API
+|========
 
 [WARNING]
 .Scripts and security
```

Comment:
Scripts are now all sandboxed, this section can be removed (cc @jdconrad @rjernst).

Reply:
++ I'll remove that admonition.